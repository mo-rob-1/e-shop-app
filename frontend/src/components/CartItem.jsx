import { Link } from "react-router-dom";

function CartItem({ item, qtyChangeHandler, removeHandler }) {
  return (
    <div>
      <div>
        <img src={item.imageUrl} alt={item.name} />
      </div>
      <Link to={`/product/${item.product}`}>
        <p>{item.name}</p>
      </Link>
      <p>${item.price}</p>
      <select value={item.qty} onChange={(e) => qtyChangeHandler(item.product, e.target.value)}>
        {[...Array(item.countInStock).keys()].map((x) => (
          <option key={x + 1} value={x + 1}>
            {x + 1}
          </option>
        ))}
      </select>
      <button type="button" onClick={() => removeHandler(item.product)}>
        <i className="fas fa-trash"></i>
      </button>
    </div>
  );
}

export default CartItem;
