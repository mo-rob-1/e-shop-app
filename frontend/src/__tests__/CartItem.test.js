import { render } from "@testing-library/react";
import CartItem from "../components/CartItem";
import { createMemoryHistory } from "history";
import { Provider } from "react-redux";
import store from "../redux/store";
import { BrowserRouter as Router } from "react-router-dom";

describe("CartItem", () => {
  it("renders the cart item", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <Provider store={store} history={history}>
        <Router>
          <CartItem />
        </Router>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders product item image", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <CartItem />
        </Router>
      </Provider>
    );

    const productItemImage = getByTestId("product-item-image");
    expect(productItemImage).toBeTruthy();
  });

  it("renders product item name", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <CartItem />
        </Router>
      </Provider>
    );

    const productItemName = getByTestId("product-item-name");
    expect(productItemName).toBeTruthy();
  });

  it("renders product item price", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <CartItem />
        </Router>
      </Provider>
    );

    const productItemPrice = getByTestId("product-item-price");
    expect(productItemPrice).toBeTruthy();
  });

  it("renders product item quantity selector", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <CartItem />
        </Router>
      </Provider>
    );

    const productItemQuantitySelector = getByTestId("product-qty-selector");
    expect(productItemQuantitySelector).toBeTruthy();
  });

  it("renders product item remove button", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <CartItem />
        </Router>
      </Provider>
    );

    const productItemRemoveButton = getByTestId("remove-item-button");
    expect(productItemRemoveButton).toBeTruthy();
  });
});
