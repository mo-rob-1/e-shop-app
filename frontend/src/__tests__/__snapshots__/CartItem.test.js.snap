// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem renders the cart item 1`] = `
<DocumentFragment>
  <div
    class="shadow rounded-tr-xl rounded-bl-xl rounded-br-xl product-item-container"
  >
    <div
      class="md:h-full"
    >
      <img
        class="rounded-tl-xl rounded-tr-xl w-full md:h-40 md:rounded-tr-none md:rounded-bl-xl object-cover"
        data-testid="product-item-image"
      />
    </div>
    <div
      class="pl-4 pr-4 pt-6 pb-6 grid gap-3"
    >
      <div
        class="flex justify-between items-center"
      >
        <div>
          <a
            href="/product/undefined"
          >
            <p
              class="font-bold mb-2"
              data-testid="product-item-name"
            />
          </a>
          <p
            class="font-bold mb-2"
            data-testid="product-item-price"
          >
            Â£
          </p>
          <div
            data-testid="product-qty-selector"
          >
            Quantity:
            <select
              class="border ml-2"
            >
              <option
                value="1"
              >
                1
              </option>
            </select>
          </div>
        </div>
        <button
          data-testid="remove-item-button"
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
